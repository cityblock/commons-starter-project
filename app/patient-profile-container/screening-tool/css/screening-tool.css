@import '../../../shared/css/sizes';

.buttonBar {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
}

.panel {
  background-color: var(--gray-background-color);
  padding: 20px var(--medium-gutter);
}

.assessmentTitle {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 10px var(--small-gutter) 10px 16px;
  border-radius: 4px;
  background-color: var(--white-color);
  margin-bottom: 10px;
  border-left: 4px solid transparent;
}

.assessmentTitle.lowRisk {
  border-left: 4px solid var(--green-color);
}

.assessmentTitle.mediumRisk {
  border-left: 4px solid var(--yellow-color);
}

.assessmentTitle.highRisk {
  border-left: 4px solid var(--red-color);
}

.disabled {
  opacity: 0.4;

  &:hover {
    cursor: not-allowed;
  }
}

.title,
.meta {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.assessment,
.error {
  height: calc(100vh - 200px);
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
}

.title {
  justify-content: flex-start;
}

.meta {
  justify-content: flex-end;
  border-left: 1px solid var(--gray-background-color);
  padding-left: 10px;
}

.titleText {
  color: var(--black-color);
  font-family: var(--roboto);
  font-weight: var(--roboto-medium-weight);
  font-size: var(--default-font-size);
  line-height: var(--default-line-height);
}

.lastUpdatedLabel {
  font-family: var(--roboto);
  font-size: var(--extra-small-font-size);
  line-height: var(--extra-small-font-line-height);
  color: var(--light-gray-font-color);
  margin-right: 5px;
}

.lastUpdatedValue {
  font-family: var(--roboto);
  font-size: var(--extra-small-font-size);
  line-height: var(--extra-small-font-line-height);
  color: var(--black-color);
}

.errorMessage {
  display: flex;
  flex-direction: column;
  width: 100%;
  border: 1px dashed var(--border-color);
  border-radius: 4px;
  padding: 50px;
  justify-content: space-between;
  align-items: center;
}

.errorMessageText {
  display: flex;
  flex-direction: row;
  width: 100%;
  align-items: center;
  justify-content: center;
}

.errorMessageLabel {
  margin-right: 5px;
}
