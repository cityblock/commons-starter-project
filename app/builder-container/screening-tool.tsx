import * as classNames from 'classnames';
import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import * as screeningToolQuery from '../graphql/queries/get-screening-tool.graphql';
/* tslint:disable:max-line-length */
import * as screeningToolEditMutation from '../graphql/queries/screening-tool-edit-mutation.graphql';
/* tslint:enable:max-line-length */
import { screeningToolEditMutationVariables, FullScreeningToolFragment } from '../graphql/types';
import * as styles from '../shared/css/two-panel-right.css';
import { IState as IAppState } from '../store';
import ScoreRangeCreateEdit from './score-range-create-edit';

interface IProps {
  routeBase: string;
  refetchScreeningTool: () => any;
  match?: {
    params: {
      screeningToolId?: string;
    };
  };
  onDelete: (screeningToolId: string) => any;
}

interface IGraphqlProps {
  screeningTool?: FullScreeningToolFragment;
  screeningToolLoading?: boolean;
  screeningToolError?: string;
  editScreeningTool: (
    options: { variables: screeningToolEditMutationVariables },
  ) => { data: { screeningToolEdit: FullScreeningToolFragment } };
}

interface IStateProps {
  screeningToolId?: string;
}

interface IState {
  deleteConfirmationInProgress: boolean;
  deleteError?: string;
  editedTitle: string;
  editingTitle: boolean;
  editTitleError?: string;
}

type allProps = IGraphqlProps & IStateProps & IProps;

export class ScreeningTool extends React.Component<allProps, IState> {
  editTitleInput: HTMLInputElement | null;
  titleBody: HTMLDivElement | null;

  constructor(props: allProps) {
    super(props);

    this.reloadScreeningTool = this.reloadScreeningTool.bind(this);
    this.onClickDelete = this.onClickDelete.bind(this);
    this.onConfirmDelete = this.onConfirmDelete.bind(this);
    this.onCancelDelete = this.onCancelDelete.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onClickToEditTitle = this.onClickToEditTitle.bind(this);
    this.focusInput = this.focusInput.bind(this);
    this.renderScoreRanges = this.renderScoreRanges.bind(this);

    this.editTitleInput = null;
    this.titleBody = null;

    this.state = {
      deleteConfirmationInProgress: false,
      deleteError: undefined,
      editedTitle: '',
      editingTitle: false,
    };
  }

  componentWillReceiveProps(nextProps: allProps) {
    const { screeningTool } = nextProps;

    if (screeningTool) {
      if (!this.props.screeningTool || this.props.screeningTool.id !== screeningTool.id) {
        this.setState(() => ({
          editedTitle: screeningTool.title,
        }));
      }
    }
  }

  reloadScreeningTool() {
    const { refetchScreeningTool } = this.props;

    if (refetchScreeningTool) {
      refetchScreeningTool();
    }
  }

  onClickDelete() {
    const { screeningToolId } = this.props;

    if (screeningToolId) {
      this.setState(() => ({ deleteConfirmationInProgress: true }));
    }
  }

  async onConfirmDelete() {
    const { onDelete, screeningToolId } = this.props;

    if (screeningToolId) {
      try {
        this.setState(() => ({ deleteError: undefined }));
        await onDelete(screeningToolId);
        this.setState(() => ({ deleteConfirmationInProgress: false }));
      } catch (err) {
        this.setState(() => ({ deleteError: err.message }));
      }
    }
  }

  onCancelDelete() {
    this.setState(() => ({ deleteError: undefined, deleteConfirmationInProgress: false }));
  }

  onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.currentTarget.value;
    const name = event.currentTarget.name;

    this.setState(() => ({ [name]: value || '' }));
  }

  async onKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
    const { screeningToolId, editScreeningTool } = this.props;
    const { editedTitle } = this.state;
    const enterPressed = event.keyCode === 13;
    const name = event.currentTarget.name;

    if (enterPressed && screeningToolId) {
      event.preventDefault();

      if (name === 'editedTitle') {
        try {
          this.setState(() => ({ editTitleError: undefined }));
          await editScreeningTool({ variables: { screeningToolId, title: editedTitle } });
          this.setState(() => ({ editTitleError: undefined, editingTitle: false }));
        } catch (err) {
          this.setState(() => ({ editTitleError: err.message }));
        }
      }
    }
  }

  onBlur(event: React.FocusEvent<HTMLInputElement>) {
    const name = event.currentTarget.name;

    if (name === 'editedTitle') {
      this.setState(() => ({ editingTitle: false }));
    }
  }

  onClickToEditTitle() {
    this.setState(() => ({ editingTitle: true }));
    setTimeout(() => (this.focusInput(this.editTitleInput), 100));
  }

  focusInput(input: HTMLInputElement | null) {
    if (input) {
      input.focus();
    }
  }

  renderScoreRanges() {
    const { screeningTool } = this.props;
    if (screeningTool && screeningTool.screeningToolScoreRanges) {
      return screeningTool.screeningToolScoreRanges.map(scoreRange => (
        <ScoreRangeCreateEdit
          key={scoreRange ? scoreRange.id : ''}
          scoreRange={scoreRange}
          screeningToolId={screeningTool.id}
        />
      ));
    }
  }

  render() {
    const { screeningTool, routeBase } = this.props;
    const {
      deleteConfirmationInProgress,
      deleteError,
      editedTitle,
      editingTitle,
      editTitleError,
    } = this.state;

    const outerContainerStyles = classNames(styles.container, {
      [styles.deleteConfirmationContainer]: deleteConfirmationInProgress,
    });
    const screeningToolContainerStyles = classNames(styles.itemContainer, {
      [styles.hidden]: deleteConfirmationInProgress,
    });
    const deleteConfirmationStyles = classNames(styles.deleteConfirmation, {
      [styles.hidden]: !deleteConfirmationInProgress,
    });
    const deleteErrorStyles = classNames(styles.deleteError, {
      [styles.hidden]: !deleteConfirmationInProgress || !deleteError,
    });
    const titleTextStyles = classNames(styles.largeText, styles.title, {
      [styles.hidden]: editingTitle,
    });
    const titleEditStyles = classNames(styles.largeTextEditor, {
      [styles.hidden]: !editingTitle,
      [styles.error]: !!editTitleError,
    });
    const closeRoute = routeBase || '/builder/tools';
    const scoreRanges = this.renderScoreRanges();

    if (screeningTool) {
      return (
        <div className={outerContainerStyles}>
          <div className={deleteConfirmationStyles}>
            <div className={styles.deleteConfirmationIcon} />
            <div className={styles.deleteConfirmationText}>
              Are you sure you want to delete this screening tool?
            </div>
            <div className={styles.deleteConfirmationButtons}>
              <div
                className={classNames(styles.deleteCancelButton, styles.invertedButton)}
                onClick={this.onCancelDelete}
              >
                Cancel
              </div>
              <div className={styles.deleteConfirmButton} onClick={this.onConfirmDelete}>
                Yes, delete
              </div>
            </div>
            <div className={deleteErrorStyles}>
              <div className={classNames(styles.redText, styles.smallText)}>
                Error deleting screening tool.
              </div>
              <div className={styles.smallText}>Please try again.</div>
            </div>
          </div>
          <div className={screeningToolContainerStyles}>
            <div>
              <div className={styles.infoRow}>
                <div className={styles.controls}>
                  <Link to={closeRoute} className={styles.close}>
                    Close
                  </Link>
                  <div className={styles.menuItem} onClick={this.onClickDelete}>
                    <div className={styles.trashIcon} />
                    <div className={styles.menuLabel}>Delete screening tool</div>
                  </div>
                </div>
              </div>
            </div>
            <div className={styles.itemBody}>
              <div className={styles.smallText}>Title:</div>
              <div
                ref={div => {
                  this.titleBody = div;
                }}
                className={titleTextStyles}
                onClick={this.onClickToEditTitle}
              >
                {screeningTool.title}
              </div>
              <div className={titleEditStyles}>
                <input
                  name="editedTitle"
                  ref={area => {
                    this.editTitleInput = area;
                  }}
                  value={editedTitle}
                  onChange={this.onChange}
                  onKeyDown={this.onKeyDown}
                  onBlur={this.onBlur}
                />
              </div>
              <div className={styles.smallText}>Risk Area:</div>
              <div className={styles.largeText}>{screeningTool.riskArea.title}</div>
              <br />
              <div className={styles.smallText}>Score Ranges:</div>
              <div>{scoreRanges}</div>
              <div className={styles.smallText}>Create score range:</div>
              <ScoreRangeCreateEdit screeningToolId={screeningTool.id} />
            </div>
          </div>
        </div>
      );
    } else {
      const { screeningToolLoading, screeningToolError } = this.props;
      if (screeningToolLoading) {
        return (
          <div className={styles.container}>
            <div className={styles.loading}>Loading...</div>
          </div>
        );
      } else if (!!screeningToolError) {
        return (
          <div className={styles.container}>
            <div className={styles.loadingError}>
              <div className={styles.loadingErrorIcon} />
              <div className={styles.loadingErrorLabel}>Unable to load screening tool</div>
              <div className={styles.loadingErrorSubheading}>
                Sorry, something went wrong. Please try again.
              </div>
              <div
                className={classNames(styles.loadingErrorButton, styles.invertedButton)}
                onClick={this.reloadScreeningTool}
              >
                Try again
              </div>
            </div>
          </div>
        );
      } else {
        return <div className={styles.container} />;
      }
    }
  }
}

function mapStateToProps(state: IAppState, ownProps: IProps): IStateProps {
  return {
    screeningToolId: ownProps.match ? ownProps.match.params.screeningToolId : undefined,
  };
}

export default compose(
  connect<IStateProps, {}, IProps>(mapStateToProps as any),
  graphql<IGraphqlProps, IProps, allProps>(screeningToolEditMutation as any, {
    name: 'editScreeningTool',
  }),
  graphql<IGraphqlProps, IProps, allProps>(screeningToolQuery as any, {
    skip: (props: IProps & IStateProps) => !props.screeningToolId,
    options: (props: IProps & IStateProps) => ({
      variables: { screeningToolId: props.screeningToolId },
    }),
    props: ({ data }) => ({
      screeningToolLoading: data ? data.loading : false,
      screeningToolError: data ? data.error : null,
      screeningTool: data ? (data as any).screeningTool : null,
      refetcScreeningTool: data ? data.refetch : null,
    }),
  }),
)(ScreeningTool);
