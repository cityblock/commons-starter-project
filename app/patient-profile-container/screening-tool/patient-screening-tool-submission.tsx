import * as React from 'react';
import { graphql } from 'react-apollo';
import { FormattedDate } from 'react-intl';
/* tslint:disable:max-line-length */
import * as patientScreeningToolSubmissionQuery from '../../graphql/queries/get-patient-screening-tool-submission-for-patient-and-screening-tool.graphql';
/* tsline:enable:max-line-length */
import { FullPatientScreeningToolSubmissionFragment } from '../../graphql/types';
import Icon from '../../shared/library/icon/icon';
import * as styles from './css/screening-tools.css';

interface IProps {
  screeningToolId: string;
  patientId: string;
}

interface IGraphqlProps {
  loading?: boolean;
  error: string | null;
  submission?: FullPatientScreeningToolSubmissionFragment;
}

type allProps = IGraphqlProps & IProps;

const PatientScreeningToolSubmission: React.StatelessComponent<allProps> = props => {
  const { submission, loading, error } = props;

  if (!loading && !error && !!submission) {
    const { score, screeningToolScoreRange, user, createdAt } = submission;
    const conductedByName = `${user.firstName} ${user.lastName}`;
    const conductedAt = (
      <FormattedDate value={createdAt} year="numeric" month="short" day="numeric" />
    );

    let scoreDescription: string = 'Unknown Score Meaning';

    if (screeningToolScoreRange) {
      scoreDescription = screeningToolScoreRange.description;
    }

    return (
      <div className={styles.section}>
        <div className={styles.container}>
          <div className={styles.titleRow}>
            <div className={styles.title}>
              <div className={styles.titleLabel}>Tool results:</div>
              <div className={styles.titleScore}>
                <div>{scoreDescription}</div>
                <div className={styles.titleScoreDivider}>-</div>
                <div>{score}</div>
                <Icon name="highlightOff" />
              </div>
            </div>
            <div className={styles.meta}>
              <div className={styles.metaGroup}>
                <div className={styles.metaLabel}>Conducted by:</div>
                <div className={styles.metaData}>{conductedByName}</div>
              </div>
              <div className={styles.metaGroup}>
                <div className={styles.metaLabel}>Date conducted:</div>
                <div className={styles.metaData}>{conductedAt}</div>
              </div>
            </div>
          </div>
          <div className={styles.summaryRow}>TBD</div>
        </div>
      </div>
    );
  } else {
    return null;
  }
};

export default graphql<IGraphqlProps, IProps, allProps>(
  patientScreeningToolSubmissionQuery as any,
  {
    skip: (props: IProps) => !props.screeningToolId,
    options: (props: IProps) => ({
      variables: {
        screeningToolId: props.screeningToolId,
        patientId: props.patientId,
        scored: true,
      },
    }),
    props: ({ data }) => ({
      loading: data ? data.loading : false,
      error: data ? data.error : null,
      submission: data
        ? (data as any).patientScreeningToolSubmissionForPatientAndScreeningTool
        : null,
    }),
  },
)(PatientScreeningToolSubmission);
