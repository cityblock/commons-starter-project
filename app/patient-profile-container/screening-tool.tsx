import * as classNames from 'classnames';
import { keys, values } from 'lodash';
import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import { connect, Dispatch } from 'react-redux';
import { push } from 'react-router-redux';
/* tslint:disable:max-line-length */
import * as patientScreeningToolSubmissionQuery from '../graphql/queries/get-patient-screening-tool-submission-for-patient-and-screening-tool.graphql';
import * as screeningToolQuestionsQuery from '../graphql/queries/get-questions.graphql';
import * as screeningToolQuery from '../graphql/queries/get-screening-tool.graphql';
import * as patientAnswersCreate from '../graphql/queries/patient-answers-create-mutation.graphql';
/* tsline:enable:max-line-length */
import {
  patientAnswersCreateMutationVariables,
  FullPatientAnswerFragment,
  FullPatientScreeningToolSubmissionFragment,
  FullQuestionFragment,
  FullScreeningToolFragment,
} from '../graphql/types';
import * as sortSearchStyles from '../shared/css/sort-search.css';
import { Popup } from '../shared/popup/popup';
import * as styles from './css/risk-areas.css';
import { PatientScreeningToolSubmission } from './patient-screening-tool-submission';
import ScreeningToolQuestion from './screening-tool-question';
import ScreeningToolResultsPopup from './screening-tool-results-popup';

interface IProps {
  screeningToolId: string;
  patientId: string;
  routeBase: string;
  patientRoute: string;
  redirectToScreeningTools?: () => any;
  screeningTool?: FullScreeningToolFragment;
  loading?: boolean;
  error?: string;
  screeningToolQuestions?: FullQuestionFragment[];
  screeningToolQuestionsLoading?: boolean;
  screeningToolQuestionsError?: string;
  createPatientAnswers?: (
    options: { variables: patientAnswersCreateMutationVariables },
  ) => { data: { patientAnswersCreate: [FullPatientAnswerFragment] } };
  refetchScreeningTool?: () => any;
  refetchScreeningToolQuestions?: () => any;
  previousPatientScreeningToolSubmissionLoading?: boolean;
  previousPatientScreeningToolSubmissionError?: string;
  previousPatientScreeningToolSubmission?: FullPatientScreeningToolSubmissionFragment;
}

export interface IQuestionsState {
  [questionId: string]: {
    answers: Array<{
      id: string;
      value: string;
    }>;
  };
}

interface IState {
  questions: IQuestionsState;
  screeningToolLoading: boolean;
  screeningToolError?: string;
  patientScreeningToolSubmissionId?: string;
}

export interface IQuestionCondition {
  id: string;
  questionId: string;
  answerId: string;
}

export class ScreeningTool extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.evaluateQuestionConditions = this.evaluateQuestionConditions.bind(this);
    this.getQuestionVisibility = this.getQuestionVisibility.bind(this);
    this.renderScreeningToolQuestion = this.renderScreeningToolQuestion.bind(this);
    this.renderScreeningToolQuestions = this.renderScreeningToolQuestions.bind(this);
    this.updateMultiSelectAnswer = this.updateMultiSelectAnswer.bind(this);
    this.updateAnswer = this.updateAnswer.bind(this);
    this.isLoading = this.isLoading.bind(this);
    this.isError = this.isError.bind(this);
    this.isLoadingOrError = this.isLoadingOrError.bind(this);
    this.onRetryLoad = this.onRetryLoad.bind(this);
    this.allQuestionsAnswered = this.allQuestionsAnswered.bind(this);
    this.renderPreviousPatientScreeningToolSubmission = this.renderPreviousPatientScreeningToolSubmission.bind(
      this,
    );

    this.state = {
      questions: {},
      screeningToolLoading: false,
    };
  }

  componentWillReceiveProps(nextProps: IProps) {
    const { screeningToolQuestions } = nextProps;
    const { questions } = this.state;

    // ScreeningTool Questions have loaded
    if (screeningToolQuestions && !this.props.screeningToolQuestions) {
      screeningToolQuestions.forEach(question => {
        if (!questions[question.id]) {
          questions[question.id] = { answers: [] };
        }
      });

      this.setState(() => ({ questions }));
    }
  }

  isLoading() {
    const { loading, screeningToolQuestionsLoading } = this.props;

    return loading || screeningToolQuestionsLoading;
  }

  isError() {
    const { error, screeningToolQuestionsError } = this.props;

    return !!error || !!screeningToolQuestionsError;
  }

  isLoadingOrError() {
    return this.isError() || this.isLoading();
  }

  allQuestionsAnswered() {
    const { screeningToolQuestions } = this.props;
    const { questions } = this.state;

    if (!screeningToolQuestions) {
      return false;
    }

    return screeningToolQuestions.every(question => {
      const questionData = questions[question.id];

      return !!questionData && !!questionData.answers.length;
    });
  }

  async onSubmit() {
    const { createPatientAnswers, patientId, screeningToolQuestions, screeningToolId } = this.props;
    const { questions } = this.state;
    const questionIds = keys(questions);

    if (createPatientAnswers && this.allQuestionsAnswered()) {
      this.setState(() => ({ screeningToolLoading: true, screeningToolError: undefined }));

      const patientAnswers = questionIds
        .map(questionId => questions[questionId].answers.map(answer => ({ ...answer, questionId })))
        .reduce((answers1, answers2) => answers1.concat(answers2))
        .map(answer => {
          const question = (screeningToolQuestions || []).find(q => q.id === answer.questionId);
          return {
            patientId,
            questionId: answer.questionId,
            applicable: this.getQuestionVisibility(question),
            answerId: answer.id,
            answerValue: answer.value,
          };
        });

      try {
        const results = await createPatientAnswers({
          variables: {
            patientId,
            patientAnswers,
            questionIds,
            screeningToolId,
          },
        });

        const resetQuestions: IQuestionsState = {};

        questionIds.forEach(questionId => {
          resetQuestions[questionId] = { answers: [] };
        });

        const { data } = results;
        const firstCreatedAnswer = data.patientAnswersCreate[0];

        this.setState(() => ({
          screeningToolLoading: false,
          screeningToolError: undefined,
          questions: resetQuestions,
          patientScreeningToolSubmissionId: firstCreatedAnswer.patientScreeningToolSubmissionId,
        }));
      } catch (err) {
        this.setState(() => ({ screeningToolLoading: false, screeningToolError: err.message }));
      }
    }
  }

  updateMultiSelectAnswer(questionId: string, answerId: string, value: string | number) {
    const { questions } = this.state;
    const questionData = questions[questionId];

    const answerIndex = questionData.answers.findIndex((answer: any) => answer.id === answerId);

    if (answerIndex > -1) {
      questionData.answers.splice(answerIndex, 1);
    } else {
      (questionData as any).answers.push({ id: answerId, value });
    }

    this.setState(() => ({
      questions: { ...questions, [questionId]: questionData },
    }));
  }

  updateAnswer(questionId: string, answerId: string, value: string | number) {
    const { questions } = this.state;
    const questionData = questions[questionId];

    (questionData as any).answers = [{ id: answerId, value }];

    this.setState(() => ({
      questions: { ...questions, [questionId]: questionData },
    }));
  }

  onChange(questionId: string, answerId: string, value: string | number) {
    const { screeningToolQuestions } = this.props;
    const fetchedQuestion = (screeningToolQuestions || []).find(
      question => question.id === questionId,
    );

    if (fetchedQuestion && fetchedQuestion.answerType === 'multiselect') {
      this.updateMultiSelectAnswer(questionId, answerId, value);
    } else if (fetchedQuestion) {
      this.updateAnswer(questionId, answerId, value);
    }
  }

  evaluateQuestionConditions(questionConditions: IQuestionCondition[]) {
    const { questions } = this.state;

    const questionAnswers = values(questions).map(questionValue => questionValue.answers);

    const conditionsStatus = {
      total: questionConditions.length,
      satisfied: 0,
    };

    if (questionAnswers.length) {
      const flattenedAnswers = questionAnswers.reduce((answers1, answers2) =>
        answers1.concat(answers2),
      );

      questionConditions.forEach(condition => {
        if (flattenedAnswers.some(answer => answer.id === condition.answerId)) {
          conditionsStatus.satisfied += 1;
        }
      });
    }

    return conditionsStatus;
  }

  getQuestionVisibility(question?: FullQuestionFragment): boolean {
    if (!question) {
      return false;
    }

    let visible: boolean = true;

    const questionConditions = question.applicableIfQuestionConditions;

    if (questionConditions && questionConditions.length) {
      const conditionsMet = this.evaluateQuestionConditions(questionConditions);

      const allTrue = conditionsMet.satisfied === conditionsMet.total;
      const oneTrue = conditionsMet.satisfied > 0;

      switch (question.applicableIfType) {
        case 'allTrue': {
          if (!allTrue) {
            visible = false;
            break;
          }
        }
        case 'oneTrue': {
          if (!oneTrue) {
            visible = false;
            break;
          }
        }
        default: {
          visible = true;
          break;
        }
      }
    }

    return visible;
  }

  renderScreeningToolQuestion(question: FullQuestionFragment, index: number) {
    const { questions } = this.state;

    const visible = this.getQuestionVisibility(question);

    return (
      <ScreeningToolQuestion
        visible={visible}
        answerData={questions[question.id]}
        onChange={this.onChange}
        key={`${question.id}-${index}`}
        question={question}
      />
    );
  }

  renderScreeningToolQuestions() {
    const { screeningToolQuestions } = this.props;

    return (screeningToolQuestions || []).map(this.renderScreeningToolQuestion);
  }

  onRetryLoad() {
    const { refetchScreeningTool, refetchScreeningToolQuestions } = this.props;

    // TODO: What to do if there is a serious error and these fuctions don't exist?

    if (refetchScreeningTool) {
      refetchScreeningTool();
    }

    if (refetchScreeningToolQuestions) {
      refetchScreeningToolQuestions();
    }
  }

  renderPreviousPatientScreeningToolSubmission() {
    const {
      previousPatientScreeningToolSubmission,
      previousPatientScreeningToolSubmissionLoading,
      previousPatientScreeningToolSubmissionError,
    } = this.props;

    return (
      <PatientScreeningToolSubmission
        submission={previousPatientScreeningToolSubmission}
        loading={previousPatientScreeningToolSubmissionLoading}
        error={previousPatientScreeningToolSubmissionError}
      />
    );
  }

  render() {
    const { screeningTool, patientRoute } = this.props;
    const { patientScreeningToolSubmissionId } = this.state;
    const title = screeningTool ? screeningTool.title : 'Loading...';

    const submitButtonStyles = classNames(styles.button, styles.saveButton, {
      [styles.disabled]: !this.allQuestionsAnswered(),
    });
    const titleStyles = classNames(styles.riskAssessmentTitle, {
      [styles.lowRisk]: false,
      [styles.mediumRisk]: true,
      [styles.highRisk]: false,
    });

    const assessmentHtml = this.isError() ? (
      <div className={styles.riskAssessmentError}>
        <div className={styles.riskAssessmentErrorMessage}>
          <div className={styles.riskAssessmentErrorIcon} />
          <div className={styles.riskAssessmentErrorMessageText}>
            <div className={styles.riskAssessmentErrorMessageLabel}>Error loading tool.</div>
            <div className={styles.riskAssessmentErrorMessageCta} onClick={this.onRetryLoad}>
              <div className={styles.riskAssessmentErrorMessageLink}>Click here to retry.</div>
              <div className={styles.riskAssessmentErrorMessageIcon} />
            </div>
          </div>
        </div>
      </div>
    ) : (
      <div className={styles.riskAssessment}>
        {this.renderPreviousPatientScreeningToolSubmission()}
        <div className={titleStyles}>
          <div className={styles.title}>
            <div className={styles.titleIcon} />
            <div className={styles.titleText}>{title}</div>
          </div>
          <div className={styles.meta}>
            <div className={styles.lastUpdatedLabel}>Last updated:</div>
            <div className={styles.lastUpdatedValue}>TBD!</div>
          </div>
        </div>
        {this.renderScreeningToolQuestions()}
      </div>
    );

    return (
      <div>
        <div className={classNames(sortSearchStyles.sortSearchBar, styles.buttonBar)}>
          <div className={submitButtonStyles} onClick={this.onSubmit}>
            Submit
          </div>
        </div>
        <div className={styles.riskAreasPanel}>{assessmentHtml}</div>
        <Popup visible={!!patientScreeningToolSubmissionId} smallPadding={true}>
          <ScreeningToolResultsPopup
            patientRoute={patientRoute}
            patientScreeningToolSubmissionId={patientScreeningToolSubmissionId}
          />
        </Popup>
      </div>
    );
  }
}

function mapDispatchToProps(dispatch: Dispatch<() => void>, ownProps: IProps): Partial<IProps> {
  return {
    redirectToScreeningTools: () => {
      dispatch(push(ownProps.routeBase));
    },
  };
}

export default compose(
  connect(undefined, mapDispatchToProps),
  graphql(screeningToolQuery as any, {
    options: (props: IProps) => ({
      variables: {
        filterType: 'screeningTool',
        filterId: props.screeningToolId,
      },
    }),
    props: ({ data }) => ({
      loading: data ? data.loading : false,
      error: data ? data.error : null,
      screeningTool: data ? (data as any).screeningTool : null,
      refetchScreeningTool: data ? data.refetch : null,
    }),
  }),
  graphql(screeningToolQuestionsQuery as any, {
    options: (props: IProps) => ({
      variables: {
        screeningToolId: props.screeningToolId,
      },
    }),
    props: ({ data }) => ({
      screeningToolQuestionsLoading: data ? data.loading : false,
      screeningToolQuestionsError: data ? data.error : null,
      screeningToolQuestions: data ? (data as any).questions : null,
      refetchScreeningToolQuestions: data ? data.refetch : null,
    }),
  }),
  graphql(patientAnswersCreate as any, { name: 'createPatientAnswers' }),
  graphql(patientScreeningToolSubmissionQuery as any, {
    skip: (props: IProps) => !props.screeningToolId,
    options: (props: IProps) => ({
      variables: {
        screeningToolId: props.screeningToolId,
        patientId: props.patientId,
      },
    }),
    props: ({ data }) => ({
      previousPatientScreeningToolSubmissionLoading: data ? data.loading : false,
      previousPatientScreeningToolSubmissionError: data ? data.error : null,
      previousPatientScreeningToolSubmission: data
        ? (data as any).patientScreeningToolSubmissionForPatientAndScreeningTool
        : null,
    }),
  }),
)(ScreeningTool);
