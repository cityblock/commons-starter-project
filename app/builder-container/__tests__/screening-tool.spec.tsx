import { shallow } from 'enzyme';
import React from 'react';
import { screeningTool } from '../../shared/util/test-data';
import { ScreeningTool as Component } from '../screening-tool';

const match = {
  params: {
    screeningToolId: screeningTool.id,
  },
};

describe('shallow rendered', () => {
  const refetchScreeningTool = jest.fn();
  const editScreeningTool = jest.fn();
  const onDelete = jest.fn();
  let instance: any;

  beforeEach(() => {
    const component = shallow(
      <Component
        match={match}
        routeBase={'/route/base'}
        screeningTool={screeningTool}
        screeningToolId={screeningTool.id}
        screeningToolLoading={false}
        screeningToolError={null}
        onDelete={onDelete}
        refetchScreeningTool={refetchScreeningTool}
        editScreeningTool={editScreeningTool}
      />,
    );
    instance = component.instance() as Component;
  });

  it('renders with screening tool', () => {
    expect(instance.render()).toMatchSnapshot();
  });

  it('confirms delete', async () => {
    await instance.onClickDelete();
    expect(instance.state.deleteConfirmationInProgress).toBeTruthy();

    await instance.onConfirmDelete();
    expect(onDelete).toBeCalledWith(screeningTool.id);
    expect(instance.state.deleteConfirmationInProgress).toBeFalsy();
  });

  it('cancels delete', async () => {
    await instance.onClickDelete();
    expect(instance.state.deleteConfirmationInProgress).toBeTruthy();

    await instance.onCancelDelete();
    expect(instance.state.deleteConfirmationInProgress).toBeFalsy();
  });

  it('handles editing', async () => {
    await instance.setState({ editedTitle: 'new title' });
    await instance.onKeyDown({
      keyCode: 13,
      currentTarget: {
        name: 'editedTitle',
      },
      preventDefault: jest.fn(),
    });
    expect(editScreeningTool).toBeCalledWith({
      variables: {
        screeningToolId: screeningTool.id,
        title: 'new title',
      },
    });
  });
});
