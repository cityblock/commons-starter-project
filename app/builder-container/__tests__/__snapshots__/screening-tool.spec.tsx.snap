// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow rendered renders with screening tool 1`] = `
<div
  className="container"
>
  <div
    className="deleteConfirmation hidden"
  >
    <div
      className="deleteConfirmationIcon"
    />
    <div
      className="deleteConfirmationText"
    >
      Are you sure you want to delete this screening tool?
    </div>
    <div
      className="deleteConfirmationButtons"
    >
      <div
        className="deleteCancelButton invertedButton"
        onClick={[Function]}
      >
        Cancel
      </div>
      <div
        className="deleteConfirmButton"
        onClick={[Function]}
      >
        Yes, delete
      </div>
    </div>
    <div
      className="deleteError hidden"
    >
      <div
        className="redText smallText"
      >
        Error deleting screening tool.
      </div>
      <div
        className="smallText"
      >
        Please try again.
      </div>
    </div>
  </div>
  <div
    className="itemContainer"
  >
    <div>
      <div
        className="infoRow"
      >
        <div
          className="controls"
        >
          <Link
            className="close"
            replace={false}
            to="/route/base"
          >
            Close
          </Link>
          <div
            className="menuItem"
            onClick={[Function]}
          >
            <div
              className="trashIcon"
            />
            <div
              className="menuLabel"
            >
              Delete screening tool
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="itemBody"
    >
      <div
        className="smallText"
      >
        Title:
      </div>
      <div
        className="largeText title"
        onClick={[Function]}
      >
        Screening Tool
      </div>
      <div
        className="largeTextEditor hidden"
      >
        <input
          name="editedTitle"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          value=""
        />
      </div>
      <div
        className="smallText"
      >
        Risk Area:
      </div>
      <div
        className="largeText"
      >
        Risk Area Title
      </div>
      <br />
      <div
        className="smallText"
      >
        Score Ranges:
      </div>
      <div />
      <div
        className="smallText"
      >
        Create score range:
      </div>
      <Apollo(Apollo(Apollo(ScoreRangeCreateEdit)))
        screeningToolId="screening-tool-id"
      />
    </div>
  </div>
</div>
`;
